Doc/add examples

mstryk opened the pull request:
Added Short examples under the Density, Cluster, Environment, and Order modules. Added doctest. 
## Description
Added 6 short examples to commonly used modules. These modules are commented and include all assumptions. These are shorter versions of the examples under https://github.com/glotzerlab/ovito-scripts. All examples have been configured to work with doctest  and will only show up if they are up to date with the API (thank you @bdice ). 

## Motivation and Context
I was tasked with helping to make the Freud website to be more user friendly towards beginners. I realized that the longer and more complex examples page was not ideal for fulfilling the simple need of referencing a computation. Thus, I suggested the addition of shortened examples so that a beginner does not have to switch to the examples page to find a couple lines of code. This is meant to be similar to how a student might search up a function on stack exchange just to see how a function is used. 

## How Has This Been Tested?
The examples were run through and passed all styling tests and were made to resemble previous examples under the Box Module.  Additionally, doctest ensures that examples are up to date with the modules in the event of an API update, or else they will simply not appear.
## Screenshots
<!-- (if appropriate) -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:
<!-- Put an `x` in all the boxes that apply. If you're unsure about any of
     these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

mstryk committed:
Added a short example of freud.Cluster.Cluster() and frued.cluster.ClusterProperties() under the compute section of the API

mstryk committed:
Added a short example of freud.density.LocalDensity under the compute section of Local Density in the API

mstryk committed:
Added a short example of freud.environment.EnvironmentCluster() under the compute section of EnvironmentCluster in the API

mstryk committed:
Added short examples of freud.order.Steinhardt, freud.order.Hexatic and frued.order.Nematic under each respective compute section in the API

mstryk committed:
removed white space

mstryk committed:
removed accidental addition of the word steinhardt

mstryk committed:
shortened additions to pass style check flake8

mstryk committed:
removed a couple words to further shorten examples ot pass flake8 style test

mstryk committed:
Update order.pyx

Added noqa 501

mstryk committed:
noqa

Added noqa E501 to Hexatic

mstryk committed:
Fixed styling/styling error

added "#" to noqa: E501

mstryk committed:
Added space before # noqa: E501

mstryk committed:
Updated Example Styling

Added Spaces before comments

mstryk committed:
Update density.pyx

Added space in Example

mstryk committed:
Update environment.pyx

Added Spaces in Examples

klywang commented on freud/cluster.pyx:
```diff
@@ -183,6 +183,18 @@ cdef class ClusterProperties(_Compute):
+            >>> # Assuming box, positions, and nlist have already been defined
+            >>> # Compute clusters using box, positions, and nlist data
+            >>> import freud
+            >>> import numpy as np
```
These can be excluded in the short examples
```suggestion

```

klywang commented on freud/cluster.pyx:
```diff
@@ -183,6 +183,18 @@ cdef class ClusterProperties(_Compute):
+            >>> import numpy as np
+            >>> cl = freud.cluster.Cluster()
+            >>> cl.compute(box, positions, neighbors=nlist)
+            >>> # Compute Cluster properties (size, position, center, etc)
```
```suggestion
            >>> # Compute cluster properties based on identified clusters
```
We may want to be a little more descriptive here. Talk about why

klywang commented on freud/density.pyx:
```diff
@@ -360,6 +360,15 @@ cdef class LocalDensity(_PairCompute):
+
+            >>> # Assuming box, positions, and system (box, points) have already been defined
+            >>> # Compute Local Density
+            >>> # r_max and diameter can be changed as needed
```
We can remove this line for the minimal example
```suggestion
            >>> # r_max and diameter can be changed as needed
```

klywang commented on freud/density.pyx:
```diff
@@ -360,6 +360,15 @@ cdef class LocalDensity(_PairCompute):
+            >>> # Assuming box, positions, and system (box, points) have already been defined
+            >>> # Compute Local Density
+            >>> # r_max and diameter can be changed as needed
+            >>> import freud
```
Remove this as well

klywang commented on freud/environment.pyx:
```diff
@@ -534,6 +534,18 @@ cdef class EnvironmentCluster(_MatchEnv):
+            >>> env_cluster = freud.environment.EnvironmentCluster()
+            >>> env_cluster.compute(system, threshold=threshold,
+            >>> neighbors={'num_neighbors': num_neighbors},
+            >>> registration=registration, global_search=global_search)
```
Pick realistic values for variables that should be a scalar or a boolean (True/False). You can just copy whatever they used from the **Example** page. Replace `system` with `(box, positions)`.

klywang commented on freud/order.pyx:
```diff
@@ -175,10 +175,24 @@ cdef class Nematic(_Compute):
+            >>> # Recommend visualizing the data with the particle_property below
+            >>> import freud
+            >>> from ovito.data import *
+            >>> import numpy as np
```
Remove these import statements.

klywang commented on freud/order.pyx:
```diff
@@ -175,10 +175,24 @@ cdef class Nematic(_Compute):
 
+        Example::
+
+            >>> # Assuming user has set director = np.array[x,y,z]
```
The synatx for `np.array[x, y, z]` should be `np.array([x, y, z])` -- but judging from the examples for the `Nematic`  order parameter, you can just use a list like `[1, 0, 0]` for the example. 

I think it would be best, though, for this example, to remove this comment and copy the value from the example for the `director` variable below.

klywang commented on freud/order.pyx:
```diff
@@ -175,10 +175,24 @@ cdef class Nematic(_Compute):
+            >>> from ovito.data import *
+            >>> import numpy as np
+            >>> order_param = freud.order.Nematic(director)
+            >>> order_param.compute(points.orientations)
```
`points` should be an `Nx3` array in these examples so they should not have an `orientation` attribute. You may want to assign fake orientations in this example or include this in your comments as a preassigned value.

klywang commented on freud/order.pyx:
```diff
@@ -175,10 +175,24 @@ cdef class Nematic(_Compute):
+            >>> order_param = freud.order.Nematic(director)
+            >>> order_param.compute(points.orientations)
+            >>> data.create_user_particle_property(name='NematicOrderParameter',
+            >>> data_type=float, data=order_param.particle_tensor
```
Remove these. These are for `Ovito`

klywang commented on freud/order.pyx:
```diff
@@ -272,6 +286,22 @@ cdef class Hexatic(_PairCompute):
+            >>> # Recommend visualizing data with the particle_property below
+            >>> import freud
+            >>> from ovito.data import *
+            >>> import numpy as np
```
Remove these import statements as well

klywang commented on freud/order.pyx:
```diff
@@ -272,6 +286,22 @@ cdef class Hexatic(_PairCompute):
+            >>> psi_k = np.copy(order_param.particle_order)
+            >>> psi_k -= np.mean(psi_k)
+            >>> data.create_user_particle_property(name='HexaticOrderParameter',
+            >>> data_type=float, data=np.angle(psi_k, deg=True)/order_param.k)
```
Remove this

klywang commented on freud/order.pyx:
```diff
@@ -272,6 +286,22 @@ cdef class Hexatic(_PairCompute):
+            >>> order_param = freud.order.Hexatic(k)
+            >>> order_param.compute(system)
+            >>> psi_k = np.copy(order_param.particle_order)
+            >>> psi_k -= np.mean(psi_k)
```
What is the purpose of including these lines in the example?
If we think it would be beneficial to show people how to access the final order parameters, it may be more useful just to print the values, like in these `hoomd` examples [[x](https://hoomd-blue.readthedocs.io/en/stable/module-hoomd-data.html)]:
```python
>>> print(order_param.particle_order)
array([ 0.2119954 -0.5681961j ,  0.18827692-0.4118587j ,
       -0.13218744-0.6692502j , ..., -0.4630672 -0.15254533j,
        0.36158592+0.08465159j,  0.15578252-0.44301566j], dtype=complex64)
```
Otherwise, it may be best just to remove these.

klywang commented on freud/order.pyx:
```diff
@@ -543,6 +573,19 @@ cdef class Steinhardt(_PairCompute):
+
+            >>> # Assuming box, points, and system (box, points) are defined
+            >>> # Recommend visualizing data with the particle_property below
+            >>> # r_max and 1 can be changed as desired
```
You can remove this

klywang commented on freud/order.pyx:
```diff
@@ -543,6 +573,19 @@ cdef class Steinhardt(_PairCompute):
+            >>> # r_max and 1 can be changed as desired
+            >>> import freud
+            >>> from ovito.data import *
+            >>> import numpy as np
```
Remove these import statements as well

klywang commented on freud/order.pyx:
```diff
@@ -543,6 +573,19 @@ cdef class Steinhardt(_PairCompute):
+            >>> ql = freud.order.Steinhardt(l=6)
+            >>> ql.compute(system, {'r_max': 3})
+            >>> ql_property = data.create_user_particle_property(name='Ql',
+            >>> data_type=float, data=ql.particle_order)
```
Remove anything related to `ovito` here

klywang reviewed (changes requested):
Thank you for doing this! Just a few changes, and then we'll be good to go:
(1) Remove import statements from the examples. If you end up using `numpy`, use the full name, rather than `np` so it's explicit. We should avoid using extra packages unless necessary for the example, however.
(2) Remove anything related to `ovito` here.
(3) Try to use realistic numbers are inputs whenever possible. You can assume the user has information like `box` or `points` already, but we want to avoid introducing too many assumed variables. It makes it harder to read. Plus, sometimes the examples are useful because they provide dummy values that give you a good idea of what that parameter should look like.

klywang commented on freud/order.pyx:
```diff
@@ -175,10 +175,24 @@ cdef class Nematic(_Compute):
+            >>> import freud
+            >>> from ovito.data import *
+            >>> import numpy as np
+            >>> order_param = freud.order.Nematic(director)
```
We may want to pick a more descriptive name for this variable, such as `nematic`? -- or `nop` from the **Examples** for consistency?

klywang commented on freud/order.pyx:
```diff
@@ -272,6 +286,22 @@ cdef class Hexatic(_PairCompute):
+            >>> import freud
+            >>> from ovito.data import *
+            >>> import numpy as np
+            >>> order_param = freud.order.Hexatic(k)
```
We may want to pick a more descriptive name here as well. Something like `hex_order` if we choose `k=6` or `k_order` for more generality.

mstryk committed:
Shortened Examples

Shortened Examples

mstryk committed:
Shortened Example

Shortened Example

bdice commented:
1. Can we try to adopt doctest or a similar tool to ensure that these examples work as a part of freud's continuous integration? Otherwise API changes can break the examples and that's really bad for users. https://docs.python.org/3/library/doctest.html

2. I would like the examples to be runnable if possible. I would avoid assuming that `system` already exists. Use the `freud.data` module to generate systems as needed (it's as short as one line to generate a random system).

mstryk committed:
Update Exa,ples

Replaced example variables with real values

mstryk committed:
Example Edits

Removed Import Statements, Changed names to reflect function, Deleted Ovito things

mstryk committed:
Shortened Examples

mstryk committed:
Update order.pyx

mstryk committed:
Changed Examples to generate box and points with Freud.data

mstryk committed:
Update order.pyx

mstryk committed:
Update order.pyx

mstryk committed:
Auto generate box and points for examples

mstryk committed:
Auto generate box and points in examples

mstryk committed:
Remove whitespace

mstryk committed:
Added Space

mstryk committed:
Removed Space

mstryk committed:
Remove whitespace

mstryk committed:
Remove whitespace

mstryk committed:
Auto generate box and positions in examples

klywang committed:
Modified to work with doctests.

bdice committed:
Add script to test docstrings in unittest.

bdice committed:
Fix some doctests.

bdice committed:
Revert long line, enabled doctest.NORMALIZE_WHITESPACE instead.

bdice committed:
Move recommendations about provided parameters into docstrings instead of doctest comment.

bdice committed:
Don't generate unnecessary box/points.

bdice committed:
Force Hexatic system to be 2D.

bdice committed:
Add return value to Steinhardt doctest.

bdice committed:
Configure doctest options.

bdice commented:
@mstryk @klywang Really nice work so far. I wasn't sure how to use `doctest` myself so I spent a half hour learning it -- I didn't want to encourage you to pursue something that was unnecessarily difficult. I pushed some commits that enable `doctest` as a part of continuous integration, and I fixed the currently failing doctests. I think it works pretty well. I hope these updates enable you to write more docstrings, now that you have working and validated examples! 👍

mstryk commented:
@bdice Thank you for adding adding doctests to the examples! I had never considered that such a thing could exist or what happens when example are outdated by the modules. I will open up the pull request from a draft to a full request. I will try to integrate doctest into any further examples I may add for other modules using your examples.

klywang reviewed (approved):
@bdice Thank you for implementing `doctest` in CI and fixing the examples! @mstryk Everything looks good now.

klywang commented:
Resolves issue #623

mstryk committed:
Update credits

mstryk committed:
Update ChangeLog

vyasr commented on doc/source/reference/credits.rst:
```diff
@@ -287,6 +287,11 @@ Mike Henry
+Michael Stryk
+
+* Added short examples into Cluster, Density, Environment, and Order Modules.
+* Updated old API in freud ovito-scripts.
```
I'd remove this line (and maybe add a credits file if one doesn't exists to the ovito-scripts repo so we can keep track of who did what there).

vyasr commented:
This is good stuff, thanks @mstryk! @bdice thanks for getting doctest working, I think we've discussed before but nice to see it in action. I think this PR is ready for a final pass from you if you'd like, otherwise we can call it ready.

vyasr committed:
Merge branch 'master' into doc/add-examples

mstryk committed:
Update Credits

Removed mention of ovito-scripts

klywang committed:
Merge branch 'master' into doc/add-examples
